engine/lib/phx/src/engine/event_bus.rs

```diff
@@ -0,0 +1,303 @@
+        Self {
+            next_tunnel_id: AtomicU32::new(0),
+            update_pass_map: HashMap::new(),
+            events: HashMap::new(),
+            max_priority_locked: false,
+        }
+    }
+
+    pub fn lock_max_priority(&mut self) {
+        self.max_priority_locked = true;
+    }
+
+    pub fn dispatch(&mut self, update_pass: UpdatePass, lua: &mut Ref<Lua>) {
+        //* Can we handle the lua state differently? */
+        //* Maybe also with future potential multiple lua states in mind */
+        let globals = lua.globals();
+        let event_tunnels: Table = globals.get("EventTunnels").expect("Unknown table");
+
+        if let Some(message_heap) = self.update_pass_map.get_mut(&update_pass) {
+            let mut message_requests: Vec<_> = message_heap.drain().collect();
+            message_requests.sort_by(|a, b| b.priority.cmp(&a.priority));
+
+            for message_request in &mut message_requests {
+                if let Some(event) = self.events.get(&message_request.name) {
+                    for subscriber in &event.subscribers {
+                        let id = subscriber.tunnel_id;
+
+                        if !message_request.completed && !message_request.stay_alive {
+                            println!(
+                                "Dispatched event {} through tunnel with id {}",
+                                message_request.name, id
+                            )
+                        }
+
+                        // todo: only dispatch to subscriber with set entity_id if defined
+                        let tunnel_func: Function = event_tunnels
+                            .get(id)
+                            .expect(&format!("Unknown tunnel with id: {}", id));
+
+                        if let Err(e) = tunnel_func.call::<_, ()>(()) {
+                            trace!("{}", e);
+                        } else {
+                            message_request.completed = true; // Mark as completed
+                        }
+                    }
+                } else {
+                    panic!("Event not found: {}", message_request.name);
+                }
+            }
+
+            for message_request in message_requests {
+                if message_request.stay_alive || !message_request.completed {
+                    message_heap.push(message_request);
+                }
+            }
+        }
+    }
+
+    pub fn dispatch_all(&mut self, mut lua: Ref<Lua>) {
+        for update_pass in UpdatePass::iter() {
+            self.dispatch(update_pass, lua.borrow_mut());
+        }
+    }
+}
+
+#[luajit_ffi_gen::luajit_ffi]
+impl EventBus {
+    pub fn register(
+        &mut self,
+        event_name: String,
+        priority: Option<u16>, //* how do i handle Options via ffi? It requires a uint16 const pointer */
+        update_pass: UpdatePass, //* how do i make the update pass enum work nicely for the lua side? Since the enum is a lua number on lua side.*/
```

Comment:
What do you mean by work nicely? It seems to be generating the right type information in the meta file. Ultimately, enums _are_ just numbers no?

Reply:
I mean on lua side for register i can not pass an UpdatePass enum since it´s a lua number. So i would have to change it to accept float32 instead and assign it to the enum in the rust side.

Reply:
I mean on lua side for register i can not pass an UpdatePass enum since it´s a lua number. So i would have to change it to accept float32 instead and assign it to the enum in the rust side.

Reply:
can i somehow make 
```lua
EventBusInstance:register("Test", nil, UpdatePass.PostFrame, nil, function() end)
```
work? So that i can pass UpdatePass.PostFrame directly. I saw some From implementation in other files but i don´t really know how to do that properly for the ffi.

Reply:
I think I understand what is the problem here. At the moment proc macro pass `UpdatePass` as the pointer here. But actually this enum is just an integer. You can tell proc macro that this type is copyable by adding it to the `COPY_TYPES` list [here](https://github.com/Limit-Theory-Redux/ltheory/blob/00a5c6475b47e86cf02d2a2099ef307237a5cf72/engine/lib/luajit-ffi-gen/src/impl_item/type_info.rs#L21).
Then proc macro won't use pointer and will pass this enum as an integer.